cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java

```diff
@@ -234,8 +238,17 @@ public void starting() {
           // If RunId is not time-based, use current time as start time
           startTimeInSeconds = TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis());
         }
-        runtimeStore.setStart(programId, runId.getId(), startTimeInSeconds, twillRunId, userArgs.asMap(),
-                              arguments.asMap());
+
+        final long finalStartTimeInSeconds = startTimeInSeconds;
+        Retries.callWithRetries(new Retries.Callable<Void, RuntimeException>() {
```

Comment:
if the method to retry doesn't throw a checked exception, you can use Retries.supplyWithRetries to make the code a little less verbose.